File: src/main/scala/com/github/kmizu/yapp/Symbol.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package com.github.kmizu.yapp

import java.util.HashMap
import java.util.Map
import java.lang.Comparable

final class Symbol private (val key: String) extends Comparable[Symbol] {
  override def toString: String = key

  override def equals(obj: Any): Boolean = super.equals(obj)

  override def hashCode(): Int = super.hashCode()

  override def compareTo(o: Symbol): Int = {
    if(this == o) return 0
    key.compareTo(o.key)
  }
}
object Symbol {
  private val cache: Map[String, Symbol] = new HashMap[String, Symbol]
  
  def intern(key: String): Symbol = {
    if(key == null) return null;
    var value = cache.get(key);
    if(value == null){
      value = new Symbol(key);
      cache.put(key, value);
    }
    value
  }
}

File: src/jp/gr/java_conf/mizu/yapp/Symbol.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package jp.gr.java_conf.mizu.yapp;

import java.util.HashMap;
import java.util.Map;

public final class Symbol implements Comparable<Symbol> {
  private static Map<String, Symbol> cache = new HashMap<String, Symbol>();
  
  public static Symbol intern(String key){
    if(key == null) return null;
    Symbol value = cache.get(key);
    if(value == null){
      value = new Symbol(key);
      cache.put(key, value);
    }
    return value;
  }
  
  private String key;
  
  private Symbol(String key){
    this.key = key;
  }
  
  public String getKey(){
    return key;
  }
  
  @Override
  public String toString() {
    return key;
  }
  
  @Override
  public boolean equals(Object obj) {
    return super.equals(obj);
  }
  
  @Override
  public int hashCode() {
    return super.hashCode();
  }

  public int compareTo(Symbol o) {
    if(this == o) return 0;
    return key.compareTo(o.key);
  }
}
